tools/build_langserver/langserver/analyzer.go

```diff
@@ -1,3 +1,199 @@
 package langserver
 
+import (
+	"core"
+	"parse"
+	"strings"
+
+	"parse/asp"
+	"tools/build_langserver/lsp"
+	"io/ioutil"
+	"fmt"
+)
+
 // TODO(bnmetrics): This file should contain functions to retrieve builtin and custom definitions of build defs
+
+// Analyzer is a wrapper around asp.Analyzer
+// This is being loaded into a handler on initialization
+type Analyzer struct {
+	parser *asp.Parser
+	BuiltIns map[string]*RuleDef
+}
+
+// RuleDef is a wrapper around asp.FuncDef,
+// it also includes a Header(function definition)
+// And Argument map stores the name and the information of the arguments this rule has
+type RuleDef struct {
+	*asp.FuncDef
+	Header string
+	ArgMap map[string]*Argument
+}
+
+// Argument is a wrapper around asp.Argument,
+// this is used to store the argument information for specific rules,
+// and it also tells you if the argument is required
+type Argument struct {
+	*asp.Argument
+	definition string
+	required bool
```

Comment:
This is already implicit in the `asp.Argument` - if `Value` is nil then there's no default and it's required.